streams/src/main/java/org/apache/kafka/streams/state/internals/PrefixedWindowKeySchemas.java

```diff
@@ -0,0 +1,405 @@
+             *     2. If timestamp is the same but key (k1) is larger than this lower range key (k2):
+             *         a. If k2 is not a prefix of k1, then k1 will always have larger byte order no
+             *            matter what seqnum k2 has
+             *         b. If k2 is a prefix of k1, since k2's seqnum is 0, after k1 appends seqnum,
+             *            it will always be larger than (k1 + seqnum).
+             */
+            return Bytes.wrap(ByteBuffer.allocate(PREFIX_SIZE + TIMESTAMP_SIZE + key.get().length)
+                .put(TIME_FIRST_PREFIX)
+                .putLong(from)
+                .put(key.get())
+                .array());
+        }
+
+        @Override
+        public Bytes lowerRangeFixedSize(final Bytes key, final long from) {
+            return TimeFirstWindowKeySchema.toStoreKeyBinary(key, Math.max(0, from),
+                0);
+        }
+
+        @Override
+        public Bytes upperRangeFixedSize(final Bytes key, final long to) {
+            return TimeFirstWindowKeySchema.toStoreKeyBinary(key, to, Integer.MAX_VALUE);
+        }
+
+        @Override
+        public long segmentTimestamp(final Bytes key) {
+            return TimeFirstWindowKeySchema.extractStoreTimestamp(key.get());
+        }
+
+        @Override
+        public HasNextCondition hasNextCondition(final Bytes binaryKeyFrom,
+            final Bytes binaryKeyTo, final long from, final long to, boolean forward) {
+            return iterator -> {
+                while (iterator.hasNext()) {
+                    final Bytes bytes = iterator.peekNextKey();
+                    final byte prefix = extractPrefix(bytes.get());
+
+                    if (prefix != TIME_FIRST_PREFIX) {
+                        return false;
+                    }
+
+                    final long time = TimeFirstWindowKeySchema.extractStoreTimestamp(bytes.get());
+
+                    // We can return false directly here since keys are sorted by time and if
+                    // we get time larger than `to`, there won't be time within range.
+                    if (forward && time > to) {
+                        return false;
+                    }
+                    if (!forward && time < from) {
+                        return false;
+                    }
+
+                    final Bytes keyBytes = Bytes.wrap(
+                        TimeFirstWindowKeySchema.extractStoreKeyBytes(bytes.get()));
+                    if ((binaryKeyFrom == null || keyBytes.compareTo(binaryKeyFrom) >= 0)
+                        && (binaryKeyTo == null || keyBytes.compareTo(binaryKeyTo) <= 0)
+                        && time >= from && time <= to) {
+                        return true;
+                    }
+                    iterator.next();
+                }
+                return false;
+            };
+        }
+
+        @Override
+        public HasNextCondition hasNextCondition(final Bytes binaryKeyFrom,
```

Comment:
That would be different since `hasNextCondition(...., true)` make it only handling forward case? How about we delete the `hasNextCondition` interface which doesn't have `forward` parameter? We can change everywhere to use `hasNextCondition` with `forward` param interface.

Reply:
I see; so it seems to me that the `forward` overloaded function is only different with the other one that its implementation has an early stop optimization at https://github.com/apache/kafka/pull/11802/files/496517aa8140a7125da117159cf28f181fb1ad34#diff-1efce76341bb9f0793d5a3fc1fdd05a016fc31db788da8d0597142fa551c4babR121-R126 while all others are the same, is that right?

Reply:
Right. We can early stop in this case.